spices/SPICE-0016-http-rewrites.adoc

```diff
@@ -0,0 +1,310 @@
+= HTTP Rewrites
+
+* Proposal: link:./SPICE-0016-http-rewrites.adoc[SPICE-0016]
+* Status: TBD
+* Implemented in: TBD
+* Category: Language, Tooling
+
+== Introduction
+
+This SPICE proposes a new evaluator option to allow rewriting HTTP URLs, with the intention of supporting package mirroring.
+
+This proposal supersedes https://github.com/apple/pkl-evolution/pull/4[SPICE-0002].
+
+NOTE: The design of the mirror server is out-of-scope for this proposal.
+
+== Motivation
+
+For systems that require run-time air-gapping, packages imported from `pkg.pkl-lang.org` or other sites on the internet are not available.
+This can be remedied by pre-populating a cache directory with all required dependencies (see link:SPICE-0001-import-graph-analyzer-api.adoc[SPICE-0001]).
+When dependencies can not be determined ahead of time, pre-populating the cache is not an option.
+Mirroring packages from internet repositories in a local store is insufficient, because the mirrored packages may have hard-coded dependencies.
+Importing mirrored packages, using URIs pointing to the mirror, will eventually evaluate `import`s of those hard-coded dependencies, referring to external sites.
+
+== Proposed Solution
+
+A new HTTP setting will be added, called `rewrites`.
+A rewrite is a mapping of HTTP prefixes.
+For example, a rewrite might look like this:
+
+.~/.pkl/settings.pkl
+[source,pkl]
+----
+amends "pkl:settings"
+
+http {
+  rewrites {
+    ["https://foo.org/"] = "https://bar.org/"
+  }
+}
+----
+
+This will replace all outbound calls to `https://foo.org/` with an outbound call to `https://bar.org/`.
+
+In practice, an HTTP rewrite for Apple's published packages might look like:
+
+.~/.pkl/settings.pkl
+[source,pkl]
+----
+amends "pkl:settings"
+
+http {
+  rewrites {
+    ["https://pkg.pkl-lang.org/"] = "https://my.internal.mirror/pkg-pkl-lang/"
+    ["https://github.com/"] = "https://my.internal.mirror/github-releases/"
+  }
+}
+----
+
+With this setup, Pkl will make HTTP calls to `https://my.internal.mirror` when downloading Pkl packages.
+In practice, this would be a server that fetches those assets from the underlying locations and responds with the same bytes.
+
+== Detailed design
+
+=== `org.pkl.core.http.RequestRewritingClient`
+
+Underneath the hood, Pkl makes all outbound requests using `org.pkl.core.http.RequestRewritingClient`.
+Currently, this client modifies outbound requests to add Pkl's user-agent, and optionally a port.
+This client will be modified to also rewrite the entire URI if set.
+
+In the case of failed requests, the thrown message includes details about the rewrite.
+
+Sample message:
+
+[source]
+----
+Pkl Error
+An I/O error occurred when making request GET /foo:
+Connection reset (request was rewritten: https://foo.com -> https://bar.com)
+----
+
+[[configuration-api]]
+=== Configuration API
+
+HTTP rewrites can be configured via the standard library, in Java as an API, as a server message API, and as a Gradle API.
```

Comment:
And via the CLI? I guess technically not an API but seems odd to make that distinction here.

Also, I think it might help to clarify how these different interfaces override each other. Looking at the implementation, it appears that any configure rewrites in a project overwrites all rewrites configured in settings.pkl. Likewise, any rewrites configured on the CLI override all rewrites configured via PklProject or settings.pkl.